### ISSUE TITLE
Philosopher cannot parse a custom fasta header.

### ISSUE THREAD

**Issue by MiguelCos** [Nesvilab/FragPipe#263]
Hello,

I am trying to run an LFQ search using IonQuant in combination with an MSFragger OpenSearch without Crystal-C but I am receiving the error in the title just after "2020-11-19 19:08:04 [INFO] - Updating Philosopher's tables..." appears in the log. 

What could be the source of this error? Please let me know if you need further information. I am attaching the log
[log_2020-11-20_09-31-57.txt](https://github.com/Nesvilab/FragPipe/files/5572276/log_2020-11-20_09-31-57.txt)

**Comment by fcyu** [comment]
Thanks for your interest in our tools. 

IonQuant doesn't support open search now. You can use mass offset search if you have a list of potential modifications and set 'Report mass shift as a variable mod' to 'Yes'.

Best,

Fengchao

**Comment by MiguelCos** [comment]
Thanks!

**Comment by fcyu** [comment]
You are welcome.

**Comment by MiguelCos** [comment]
Hello,

I just ran a new test with the same data, closed  search + IonQuant, no cristal-C. The same error appears. 

It seems to stop at this point 
````
2020-11-20 21:58:06 [INFO] - Updating Philosopher's tables...
2020-11-20 21:58:06 [ERROR] - null
```

Any clue what might be happening? Let me know if you need further info.


[log_2020-11-20_22-08-42_closed_search_ionquant.txt](https://github.com/Nesvilab/FragPipe/files/5576131/log_2020-11-20_22-08-42_closed_search_ionquant.txt)

**Comment by fcyu** [comment]
Hmm, I have no clue now. Could you please send us your raw files and fasta files? I need to reproduce the error. We will keep it confidential. I can give you a link to upload if it is appropriate.

Thanks,

Fengchao

**Comment by MiguelCos** [comment]
Many thanks for taking a look!

I would gladly send a link with the data to the email on your GitHub page if you don't mind.

Best,
MIguel

**Comment by fcyu** [comment]
Hi Mlguel,

I received your file and reproduced the error. It turns out that most of the entries of `Protein` column in `psm.tsv` are empty:
![image](https://user-images.githubusercontent.com/6926299/99865091-7f30ba80-2b75-11eb-9550-809b2e6a360c.png)

I guess this is because Philosopher cannot parse your fasta file. Following is an example of your fasta file's header.
```
>mz|transcript:ENST00000467637_D152A|transcript:ENST00000467637 variant:19\t35500540\t.\tT\tG\t1099.03\t.\tANN=G|missense_variant|MODERATE|ENSE00001846640|GENE_ENSE00001846640|transcript|transcript:ENST00000467637|protein_coding|9/13|c.455A>C|p.Asp152Ala|474/827|455/606|152/201||\tGT:AD:DP:GQ:PL\t1/1:0,38:38:99:1113,112,0  variant:19\t35500540\t.\tT\tG\t1099.03\t.\tANN=G|missense_variant|MODERATE|ENSE00001846640|GENE_ENSE00001846640|transcript|transcript:ENST00000467637|protein_coding|9/13|c.455A>C|p.Asp152Ala|474/827|455/606|152/201||\tGT:AD:DP:GQ:PL\t1/1:0,38:38:99:1113,112,0 OS=Homo sapiens GN=ENSE00001846640
```

@prvst Could you please take a look and make the parser more robust when you have time? 

Thanks,

Fengchao

**Comment by MiguelCos** [comment]
I see! Many thanks for taking a look! I might as well do some tweaking on the fasta file to make readable by philosopher. This is a proteogenomics database as you might have noticed, so I didn't want to modify the header and loose some info.

**Comment by prvst** [comment]
Hi @MiguelCos, please read our [documentation](https://github.com/Nesvilab/philosopher/wiki/How-to-Prepare-a-Protein-Database) on how to format and prepare your FASTA file to be used with Philosopher.

**Comment by MiguelCos** [comment]
Hello @prvst, thanks! I will prep my fasta file accordingly

**Comment by acesnik** [comment]
@MiguelCos Were you able to figure this out? It looks like you're using a Spritz database, which is cool! I'd like to see how you adapted the header and maybe integrate that into Spritz.

**Comment by MiguelCos** [comment]
Hey @acesnik,  I hope to have a clean R script within the next few days that I can share with you. It would be amazing if this kind of FragPipe-ready header can be part of Spritz.

I will now open an Issue in the Spritz repo to discuss some of the ideas that I have been struggling with for the 'reformating' of the headers and the fasta in general.

**Comment by fcyu** [comment]
Hi Felipe @prvst ,

Is it possible to make Philosopher support Spritz's protein header?

Thanks,

Fengchao

**Comment by prvst** [comment]
Yes, we can add more formats. Please send me a FASTA file as a reference, and also a PeptideProphet pep.xml file, so I can see how the prophets parse the header.

**Comment by prvst** [comment]
@MiguelCos Philosopher will follow the parsing patterns from PeptideProphet, and ProteinProphet to annotate some of those fields. Some, like the Gene name, the Organism name, and even the Protein ID, are marked with specific tags that help our program to find this information. Your database header seems to contain several in-house or non-informative tags. Because of that, even the Prophets avoid parsing the header, and just consume everything from it, which is quite long, and uninformative, see below an example of how PeptideProphet reads, and uses the information from the database:

![image](https://user-images.githubusercontent.com/403705/131352601-95ad49b4-cbf3-4285-8ba8-ddeae9a9214c.png)

Even though in theory we could tell Philosopher how to read it, it would be impossible from our end to make the Prophets do the same, since they are not maintained by us. The ideal scenario would be making the programs to support that format, but since the headers are very long, and the tools are just ignoring the format, I suggest you use the script above to put it in a more understandable format for your analysis. Have in mind that the super long header will have to be in the reports as well, so it's better to have the information parsed before using it.

**Comment by fcyu** [comment]
It seems to me that PeptideProphet uses everything before the first space as the protein ID and rest as the protein description, which is the best approach for this database. I guess the xml files would be bigger than those from using UniProt fasta. But what is the issue preventing Philosopher from supporting this format?

Best,

Fengchao

**Comment by guoci** [comment]
Has this been resolved? Or are we waiting for a reply from @prvst ?

**Comment by acesnik** [comment]
From my understanding, it's not ideal to parse long custom fasta headers. 

I'm planning to implement a change to Spritz that makes a short, unique place-holder as the header and an accompanying lookup table with the additional information.

In the meantime, @MiguelCos made a script to do that post-hoc, which he posted above.
